pkg/tsdb/elasticsearch/elasticsearch.go

```diff
@@ -2,50 +2,113 @@ package elasticsearch
+	eslog.Debug("initializing")
+	im := datasource.NewInstanceManager(NewInstanceSettings())
+	factory := coreplugin.New(backend.ServeOpts{
+		QueryDataHandler: newExecutor(im, s.Cfg, s.HTTPClientProvider),
+	})
+	if err := s.BackendPluginManager.Register("elasticsearch", factory); err != nil {
+		eslog.Error("Failed to register plugin", "error", err)
+	}
+	return nil
+}
+
+// executor represents a handler for handling elasticsearch datasource request
+type executor struct {
 	httpClientProvider httpclient.Provider
 	intervalCalculator interval.Calculator
+	im                 instancemgmt.InstanceManager
+	cfg                *setting.Cfg
 }
 
-// New creates a new Executor func.
+// newExecutor creates a new executor func.
 // nolint:staticcheck // plugins.DataPlugin deprecated
-func New(httpClientProvider httpclient.Provider) func(*models.DataSource) (plugins.DataPlugin, error) {
+func newExecutor(im instancemgmt.InstanceManager, cfg *setting.Cfg, httpClientProvider httpclient.Provider) *executor {
 	// nolint:staticcheck // plugins.DataPlugin deprecated
-	return func(dsInfo *models.DataSource) (plugins.DataPlugin, error) {
-		return &Executor{
-			httpClientProvider: httpClientProvider,
-			intervalCalculator: interval.NewCalculator(),
-		}, nil
+	return &executor{
+		im:                 im,
+		cfg:                cfg,
+		httpClientProvider: httpClientProvider,
 	}
 }
 
-// Query handles an elasticsearch datasource request
-//nolint: staticcheck // plugins.DataResponse deprecated
-func (e *Executor) DataQuery(ctx context.Context, dsInfo *models.DataSource,
-	tsdbQuery plugins.DataQuery) (plugins.DataResponse, error) {
-	if len(tsdbQuery.Queries) == 0 {
-		return plugins.DataResponse{}, fmt.Errorf("query contains no queries")
+func (e *executor) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	if len(req.Queries) == 0 {
+		return &backend.QueryDataResponse{}, fmt.Errorf("query contains no queries")
 	}
 
-	client, err := es.NewClient(ctx, e.httpClientProvider, dsInfo, *tsdbQuery.TimeRange)
+	dsInfo, err := e.getDSInfo(req.PluginContext)
 	if err != nil {
-		return plugins.DataResponse{}, err
+		return &backend.QueryDataResponse{}, err
 	}
 
-	if tsdbQuery.Debug {
-		client.EnableDebug()
+	client, err := es.NewClient(ctx, e.httpClientProvider, dsInfo, req.Queries[0].TimeRange)
+	if err != nil {
+		return &backend.QueryDataResponse{}, err
 	}
 
-	query := newTimeSeriesQuery(client, tsdbQuery, e.intervalCalculator)
+	query := newTimeSeriesQuery(client, req.Queries, e.intervalCalculator)
 	return query.execute()
 }
+
+func NewInstanceSettings() datasource.InstanceFactoryFunc {
+	return func(settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+		jsonData := map[string]interface{}{}
+		err := json.Unmarshal(settings.JSONData, &jsonData)
+		if err != nil {
+			return nil, fmt.Errorf("error reading settings: %w", err)
+		}
+		model := models.DataSource{
```

Comment:
Would replace this with a specific typed struct for the elasticsearch data source, others like cloudwatch, opentsdb and graphite have used a struct named `datasourceInfo`. For the client package which takes `models.DataSource` as argument I would define a typed struct for that in the client package.

Reply:
Yeah, thought about that, but also thought that it would maybe be a good practice to try and share code, since some fields in `models.DataSource` are common between datasources. WDYT?